### PR TITLE
Updates from GMAO

### PR DESCRIPTION
**Pull request by lizziel**
This PR brings in changes made to HEMCO for use in [GEOSgcm](https://github.com/GEOS-ESM/GEOSgcm) with GEOS-Chem 13.3.4. 

Changes include:
1. Skip check for `PBLHEIGHT` field if not running GEOS. There are HEMCO applications in GEOS that don't require `PBLHEIGHT` to be defined.
2. Change criteria for assuming restart field is all fill values. The new criteria is max value is negative. Previously it was global sum equal to zero.
3. Force the soil NOx code to grab the restart fields from the internal state on each call if using GEOS, rather than rely on whatever is in local memory.
4. Rework how to determine when to read volcano table. The new method is to keep track of the emissions data day that is in memory (new variable `YmdOnFile`) and then compare the current simulation day against that.
5.  Edits to prevent undefined variable when calculating vertical scale factor. This updates adds handling for when `HcoState%Grid%BXHEIGHT_M%Val` is associated and `HcoState%Options%VertWeight` is false.

### CODE REVIEW

**Comment by christophkeller** [comment]
> No need to include the new line in CMakeLists.txt that was added for the 3km test runs.

File: src/Core/hco_calc_mod.F90 [hunk 1]
```diff
@@ -2679,25 +2687,73 @@ SUBROUTINE GetDilFact( HcoState,   EmisL1, EmisL1Unit, EmisL2,             &
...
+          ENDDO
+
+          ! Dilution factor using fractional levels
+          IF ( UppLLR <= LowLLR ) THEN
+             DilFact = 1.0_hp / REAL(UppLL-LowLL+1,hp)
+          ELSE
+             DilFact = 1.0_hp / (UppLLR-LowLLR)
+          ENDIF
+
+#else
```
Comment (christophkeller) [inline src/Core/hco_calc_mod.F90:2752]: Adding the error message here is not correct. To maintain the old behavior (outside of MODEL_GEOS), it should be:
`DilFact = 1.0_hp / REAL(UppLL-LowLL+1,hp)`

However, I'd recommend to add this update to the standard code, rather than using a MODEL_GEOS bracket. The new code is more accurate for cases where you specify fixed emission depths (e.g., if you want to emit everything into the bottom 5000 meters).

**Review by christophkeller — CHANGES_REQUESTED** [review]
There was a merge error in `hco_calc_mod.F90`, see my comment there. Everything looks good otherwise.

File: src/Core/hco_calc_mod.F90 (hunk 1 above)
Comment (lizziel) [reply src/Core/hco_calc_mod.F90:2752]: Got it. I removed the message that crept in. I also removed the `MODEL_GEOS` ifdef so that the new handling for when `HcoState%Grid%BXHEIGHT_M%Val` is associated and `HcoState%Options%VertWeight `is false becomes standard.

File: src/Core/hco_calc_mod.F90 [hunk 2]
```diff
@@ -904,7 +906,11 @@ SUBROUTINE Get_Current_Emissions( HcoState, BaseDct,   nI,   nJ,           &
           RETURN
        ENDIF
     ELSE
+#ifdef MODEL_GEOS
+       LevDct1_Unit = HCO_EMISL_LEV
```
Comment (lizziel) [inline src/Core/hco_calc_mod.F90:910]: Could you give background on why this is needed for GEOS but not GC-Classic/GCHP.

File: src/Extensions/hcox_soilnox_mod.F90 [hunk 3]
```diff
@@ -415,7 +415,11 @@ SUBROUTINE HCOX_SoilNOx_Run( ExtState, HcoState, RC )
     ! after the rewinding the clock.
     !---------------------------------------------------------------
 
+#if !defined ( MODEL_GEOS ) 
     IF ( FIRST .OR. HcoClock_Rewind( HcoState%Clock, .TRUE. ) ) THEN
+#else
+    IF ( .TRUE. ) THEN
```
Comment (lizziel) [inline src/Extensions/hcox_soilnox_mod.F90:421]: Same question here as earlier - could you give a brief summary on why this change is needed for GEOS?

File: src/Extensions/hcox_volcano_mod.F90 [hunk 4]
```diff
@@ -563,18 +565,20 @@ SUBROUTINE ReadVolcTable( HcoState, ExtState, Inst, RC )
...
 
-    ! Do only if it's a new day...
-    IF ( HcoClock_NewDay( HcoState%Clock, EmisTime=.TRUE. ) ) THEN
-
-       ! Get current year, month, day
-       CALL HcoClock_Get ( HcoState%Clock, cYYYY=YYYY, cMM=MM, cDD=DD, RC=RC )
-       IF ( RC /= HCO_SUCCESS ) RETURN
-
+    ! Get current year, month, day
+    CALL HcoClock_Get ( HcoState%Clock, cYYYY=YYYY, cMM=MM, cDD=DD, RC=RC )
```
Comment (lizziel) [inline src/Extensions/hcox_volcano_mod.F90:569]: Could you give a brief explanation on why this change is here?

File: src/Core/hco_calc_mod.F90 (hunk 2 above)
Comment (christophkeller) [reply src/Core/hco_calc_mod.F90:947]: The first bracket is simply to make HEMCO work in an ESMF environment even if the PBLHEIGHT field is not defined. There are HEMCO applications in GEOS that don't require PBLHEIGHT to be defined.

The second update is to make sure that variable LevDct1_Unit is always defined. This should also go into the standard code, I think

File: src/Extensions/hcox_volcano_mod.F90 (hunk 4 above)
Comment (christophkeller) [reply src/Extensions/hcox_volcano_mod.F90:593]: This change was motivated by the same issue of calling `SetHcoTime` twice. This had the effect that `HcoClock_NewDay( HcoState%Clock, EmisTime=.TRUE. )` would always return .FALSE. and the volcano emissions would thus never be updated. I changed the code so that it now keeps track of the emissions data day that is in memory (variable `YmdOnFile`) and then compares the current simulation day against that. This is much cleaner, in my opinion, and works for all situations (first time step, rewinding the clock, etc.).

**Review by christophkeller — APPROVED** [review]
Looks good now.

File: src/Core/hco_calc_mod.F90 (hunk 2 above)
Comment (lizziel) [reply src/Core/hco_calc_mod.F90]: Okay, sounds good.

File: src/Core/hco_restart_mod.F90 [hunk 5]
```diff
@@ -330,9 +330,9 @@ SUBROUTINE HCO_RestartGet_3D( HcoState, Name,   Arr3D, &
         RETURN
     ENDIF
 
-    ! If field is all zero assume it to be not filled
+    ! If field is all negatives assume it to be not filled
     IF ( FLD ) THEN
```
Comment (lizziel) [inline src/Core/hco_restart_mod.F90:334]: Hi @christophkeller, we ran into an issue in the GCHP benchmark which Bob and I traced to this code. It went under the radar before since I was unaware we used hco_restart_mod in GCHP. Is this change supposed to correspond to a MAPL version update? As is, if a restart variable is not in the GCHP restart file it will be bootstrapped to zero. Is it now set to negative values?

Comment (lizziel) [reply src/Core/hco_restart_mod.F90:334]: I am editing this to include all zeros as well as all negatives in the criteria for restart field not filled. Let me know if that's an issue for GEOS.

Comment (christophkeller) [reply src/Core/hco_restart_mod.F90:334]: Checking for all zeros is dangerous because some restart files are zero in most places except for a few. Since this check is performed on each CPU individually, it can then quite easily happen that the code accidentally assumes a restart field to be undefined over some CPUs (but not others). The check for all negatives is safer in that regard (albeit also not ideal). Setting it back to all zero might work for GCHP but will be problematic for GEOS.

One option for bootstrapping a restart field is to create a dummy field in your gchp_internal_rst with the variable name of interest and with all values set to -999. This is how I usually bootstrap fields for GEOS.

Comment (lizziel) [reply src/Core/hco_restart_mod.F90:334]: Okay, sounds good. I'll to put the new code in MODEL_GEOS ifdefs as a quick fix for 13.4.0 release and then work on a better solution for 14.0.
